drag_and_drop_v2/public/js/drag_and_drop.js

```diff
@@ -1917,17 +1931,40 @@ function DragAndDropBlock(runtime, element, configuration) {
     };
 
     var isPastDue = function () {
-        return !configuration.has_deadline_passed;
+        return configuration.has_deadline_passed;
     };
-
-    var canShowAnswer = function() {
-        return configuration.mode === DragAndDropBlock.ASSESSMENT_MODE && !attemptsRemain();
+    var isAttempted = function () {
+        return state.attempts > 0;
+    };
+    var isCorrect = function () {
+        return state.last_action_correct;
+    };
+    var isClosed = function () {
+        if(isPastDue() || !attemptsRemain()) return true;
+        return false;
     };
 
     var attemptsRemain = function() {
         return !configuration.max_attempts || configuration.max_attempts > state.attempts;
     };
 
+    var canShowAnswer = function() {
+        if(configuration.mode !== DragAndDropBlock.ASSESSMENT_MODE) return false;
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.NEVER) return false;
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.ATTEMPTED) return isAttempted() || isPastDue();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.ANSWERED) return isCorrect();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.CLOSED) return isClosed();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.FINISHED) return isClosed() || isCorrect();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.CORRECT_OR_PAST_DUE) return isCorrect() || isPastDue();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.PAST_DUE) return isPastDue();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.ALWAYS) return true;
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.AFTER_ALL_ATTEMPTS) return !attemptsRemain();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.AFTER_ALL_ATTEMPTS_OR_CORRECT) return !attemptsRemain() || isCorrect();
+        else if (configuration.showanswer === SHOW_ANSWER_STATUSES.ATTEMPTED_NO_PAST_DUE) return isAttempted();
+  
+        return false;
+    };
```

Comment:
- ```"showanswer": self.is_answer_available``` will be used only at question load when calling ```student_view_data```, right?

My understanding is that the Show answer button will not work after ```do_attempt```

- Also ```showanswer``` is a [course-setting](https://github.com/abodacs/xblock-drag-and-drop-v2/blob/ca61d3902bb3af60b2970bdf8f0c4a6c36ba773a/drag_and_drop_v2/drag_and_drop_v2.py#L135)  and passing it as a boolean I find confusing.

But I agree with you unifying the logic should be the goal

Reply:
@Agrendalath I requested this addition as the client state for the "Show Answer" button doesn't get updated after user interaction. For example, when set in the "answered" mode, submitting the answer doesn't enable it without reload.

Reply:
@abodacs, @tecoholic, `do_attempt` is returning a dict with updated XBlock state. What would prevent us from returning `"showanswer": self.is_answer_available` [here](https://github.com/openedx/xblock-drag-and-drop-v2/blob/3fddeb57ce28c98d2ad40db6ecb074a69b9e2a0f/drag_and_drop_v2/drag_and_drop_v2.py#L571-L578) and updating it [here](https://github.com/openedx/xblock-drag-and-drop-v2/blob/3fddeb57ce28c98d2ad40db6ecb074a69b9e2a0f/drag_and_drop_v2/public/js/drag_and_drop.js#L1868-L1897)?

We definitely do not want to implement this logic in two places (Python and JS) because it's hard to maintain it and keep it in sync.

Reply:
@Agrendalath Ah. That would be a more appropriate solution. :100: I agree about not duplicating the logic in 2 places. 
cc: @abodacs

Reply:
@Agrendalath

I added `"answer_available": self.is_answer_available` becuase `showanswer` is a [course-setting](https://github.com/abodacs/xblock-drag-and-drop-v2/blob/ca61d3902bb3af60b2970bdf8f0c4a6c36ba773a/drag_and_drop_v2/drag_and_drop_v2.py#L135) and passing it as a boolean causes tests to fail.

Thanks

cc: @tecoholic